data/kernelstub.1
.TH KERNELSTUB "1"
.\" To view this file while editing, run it through groff:
.\"   groff -Tascii -man kernelstub.1 | less

.SH NAME
kernelstub \- The automatic Linux kernel EFIstub manager
.SH SYNOPSIS
.B kernelstub
[
.B \-e
.I esp-path
]
[
.B \-r
.I root-fs-path
]
[
.B \-k
.I kernel-image-path
]
.br
           [
.B \-i
.I initrd-image-path
]
[
.B \-o 
.I "kernel-options"
]
.br
           [
.B \-a
.I "kernel-options"
]
[
.B \-d
.I "kernel-options"
]
.br
           [
.B \-g
.I log-file-path
]
[
.B \-h
]
[
.B \-p
]
[
.B \-f
]
[
.B \-l
|
.B \-n
]
.br
           [
.B \-s
|
.B \-m
]
[
.B \-v...
]

.SH DESCRIPTION
Kernelstub is a program to setup and configure booting without a traditional 
bootloader.
It can configure booting through systemd-boot compatible loader files, or 
through the kernel's built-in efi stub loader.
It also runs automatically when the kernel is updated to keep this 
configuration up to date.
.br
It supports adding/setting different kernel command line options as well as 
keeping older kernels available for use as a backup.
It can operate on an ESP and root partition which are different from the one 
the system is currently booted from.
.SH COMMAND LINE OPTIONS
.TP
.BI "\-e, --esp-path " path
Manually setting the path to the EFI System Partition. 
This value is saved into the configuration. 
Default is /boot/efi.
.TP
.BI "\-r, --root-path " path
Manually specify the path to the root partition, in case kernelstub is being 
run on a system other than the target.
Defaults to /.
.TP
.BI "\-k, --kernel-path " path
Specify the location of the Linux kernel image/vmlinuz.
Defaults to /vmlinuz.
.TP
.BI "\-i, --initrd-path " path
Specify the location of the initrd image.
Defaults to /inird.img.
.TP
.BI "\-o, --options " 'options'
Specify the kernel cmdline options/boot flags.
This will overwrite any existing options in the configuration.
Use single or double quotes to surround multiple options.
This option will save the specified changes into the configuration.
.br
If you want to add/remove options from the existing 
configuration, see 
.B \-a
and 
.B \-d
.TP
.BI "\-a, --add-options " 'options'
Add kernel cmdliing options into the existing configuration.
This will avoid adding duplicate items if they're already present. 
Use single or double quotes to surround multiple options.
This option will save the specified changes into the configuration.
.TP
.BI "\-d, --delete-options " 'options'
Remove existing kernel cmdline options from the existing configuration.
If an option specified here is not present in the list of options, kernelstub
will silently ignore it.
Use single or double quotes to surround multiple options.
This option will save the specified changes into the configuration.
.TP
.BI "\-g, --log-file " 'path'
Specify an alternative log file location. 
.TP
.B \-h, --help 
Prints the usage information and exits. 
.TP
.B \-p, --print-config
Prints the current configuration settings and exits. 
.TP
.B \-f, --force-update
Forcibly update the system loader.conf file to set the current OS as the 
default.
This may change your system default boot order. 
.TP
.B \-l, --loader 
Creates a systemd-boot compatible loader entry in ESP/loader/entries for the
current OS.
This option is saved in the configuration.
.TP
.B \-n, --no-loader.
Don't create a loader entry for the current OS. 
This option disables the behavior of 
.B \-l
and is saved in the configuration.
.TP
.B \-s, --stub 
Enables automatic management of the kernel efistub bootloader. 
This option is saved in the configuration.
.TP
.B \-m, --manage-only 
Disables automatic management of the kernel efistub bootloader.
This option disables the behavior of 
.B \-s
and is saved in the configuration. 
.TP
.B \-v, --verbose
Make program output more verbose. 
Up to two 
.B \-v
flags can be used at once (additional flags are ignored).
.TP
.SH FILES 
.IP \fI/etc/kernelstub/configuration\fP
Default location of the kernelstub configuration file.
The file is a JSON format file with two main configurations in it;
.I 'default'
and
.I 'user'.
The 'esp_path' key is a string, 'config_rev' is an int, 'kernel_options' is a
list of strings, and all other keys are booleans.
.br
It is highly recommended to use the kernelstub utility to modify the 
configuration rather than by editing the configuration file directly.
See the 
.B CONFIGURATION
section for more details.
.PP
.IP \fI/etc/default/kernelstub\fP
This is a vendor-supplied file that can contain certain options for individual
OSs or hardware-specific values. 
.SH CONFIGURATION
Specific configuration defaults may have been modified by your OS developer or
hardware vendor. 
.br
Each kernelstub configuration contains the following keys:
.IP \fIkernel_options\fP
This is a list of strings, with each string being an individual kernel cmdline
option.
.br
Default: ["quiet", "splash"]
.br
Configured using the 
.B \-a, -d, 
and 
.B -o
flags.
.PP
.IP \fIesp_path\fP
String - Points to the path where the EFI System Partition is mounted.
.br
Default: "/boot/efi".
.br
Configured using the 
.B \-e
flag. 
.PP
.IP \fIsetup_loader\fP
boolean - enables or disables installing the loader entry file.
.br
.I false:
(default) Skips installing a loader entry file.
.br
.I true:
Installs a loader entry file.
.br
Configured using the 
.B \-l
/
.B \-n
flags. 
.PP
.IP \fImanage_mode\fP
boolean - toggles between installing the efistub bootloader or using 
management-only mode.
.br
.I false:
(default) sets up the Linux kernel built-in efistub bootloader in the system
NVRAM.
.br
.I true:
Skips setting up the built-in efistub bootloader.
.br
Configured using the 
.B \-s
/
.B \-m
flags. 
.PP
.IP \fIforce_update\fP
boolean - Forcibly overwrites the main systemd-boot configuration on each 
update.
.br
.I false:
(default) Does not automatically modify the systemd-boot configuration to make 
the current OS the default.
.br
.I true:
Overwrites the systemd-boot configuration on each update to ensure the current
OS is the default.
.br
This option cannot be enabled from the command line and must be enabled in the
configuration file directly. 
This is due to its ability cause the system to lose alternate boot entries. 
.PP
.IP \fIlive_mode\fP
boolean - Live mode allows updates on run on the live system without triggering
kernelstub. 
When live mode is enabled, kernelstub silently exits successfully
to allow software updates to work without overwriting the current boot 
configuration.
If kernelstub is run manually, live mode will be automatically disabled. 
.I false:
(default) Disables live mode.
.br
.I true:
Enables live mode.
.PP
.IP \fIconfig_rev\fP
integer - Tells kernelstub what format of configuration to expect. 
.br
If this value is lower than the current configuration revision supported by the
code, kernelstub will attempt to automatically migrate the configuration to the
new version. 
.PP
.SH BUGS
Please report bugs to https://github.com/isantop/kernelstub/issues
.SH EXAMPLE
To set up the kernel efistub bootloader to be the default boot option
.PP
.RS 
\f(CWsudo kernelstub\fP
.RE
.PP
To include some output
.PP
.RS 
\f(CWsudo kernelstub \-v\fP
.RE
.PP
To use kernelstub as a manager for systemd-boot configurations
.PP
.RS 
\f(CWsudo kernelstub \-vlm\fP
.RE
Note that the l and m flags are only required once; they are saved in the 
configuration file.
.PP
To add the "quiet" kernel option and remove the "splash" option:
.PP
.RS 
\f(CWsudo kernelstub \-a 'quiet' -d 'splash'\fP
.RE
.PP
If you have lost your boot configuration because another OS overwrote your 
setup, you can recover like so
.PP
.RS 
\f(CWsudo mount /dev/root_partition /mnt\fP
.br
\f(CWsudo mount /dev/esp_partition /mnt/boot/efi\fP
.br
\f(CWsudo kernelstub \\\fP
\f(CW    \--root-partition /mnt \\\fP
\f(CW    \--esp-path /mnt/boot/efi \\\fP
\f(CW    \--kernel-path /mnt/vmlinuz \\\fP
\f(CW    \--initrd-path /mnt/initrd.img \\\fP
\f(CW    \--options 'quiet splash' \\\fP
\f(CW    \-vslf\fP
.RE
.PP
Adjust your mount commands to correctly mount your root and ESP partitions.
.SH AUTHOR
Ian Santopietro <isantop@gmail.com>
.SH INTERNET RESOURCES
Main website/git repository:  https://github.com/isantop/kernelstub
.br
.SH SEE ALSO
efiboomgr(8), systemd-boot(7)


debian/changelog
kernelstub (3.2.0) disco; urgency=medium

  * Reformatted output
  * Revisions to make code cleaner and more pythonic
  * Adds fix for buggy UUID detection code (#22)
  * Adds UUID and hostname to entries for better organization (#23)

 -- Ian Santopietro <isantop@gmail.com>  Fri, 17 May 2019 11:43:49 -0600

kernelstub (3.1.0) bionic; urgency=medium

  * Add logging to systemd journald

 -- Ian Santopietro <ian@system76.com>  Tue, 25 Sep 2018 10:53:23 -0600

kernelstub (3.0.0) bionic; urgency=medium

  * Breaking API Change in 2.3.0
    Existing option renamed/removed: --dry-run 
    Short-form changed

 -- Ian Santopietro <ian@system76.com>  Wed, 13 Jun 2018 10:39:08 -0600

kernelstub (2.3.0) bionic; urgency=medium

  * Add flag to remove kernel options from the configuration

 -- Ian Santopietro <ian@system76.com>  Wed, 13 Jun 2018 10:03:03 -0600

kernelstub (2.2.1) bionic; urgency=medium

  * Cause kernelstub to exit without an error code if the initrd or kernel
    images can't be found.

 -- Ian Santopietro <ian@system76.com>  Wed, 23 May 2018 14:14:34 -0600

kernelstub (2.2.0) bionic; urgency=medium

  * Added an option to add kernel options

 -- Ian Santopietro <ian@system76.com>  Tue, 01 May 2018 10:15:52 -0700

kernelstub (2.1.0) bionic; urgency=medium

  * New live-mode feature to prevent automatic runs on live disks

 -- Ian Santopietro <ian@system76.com>  Tue, 17 Apr 2018 10:52:20 -0600

kernelstub (2.0.8) artful; urgency=medium

  * Pyflakes in build process
  * Code review feedback
  * More reliable UUID detection
  * Less embedded sh

 -- Ian Santopietro <ian@system76.com>  Fri, 23 Feb 2018 10:43:42 -0700

kernelstub (2.0.7) artful; urgency=medium

  * Rewrite drive detection code to be more reliable

 -- Ian Santopietro <ian@system76.com>  Thu, 15 Feb 2018 12:17:13 -0700

kernelstub (2.0.6.1) artful; urgency=medium

  * Fix cmdline installation order

 -- Ian Santopietro <ian@system76.com>  Wed, 14 Feb 2018 11:15:08 -0700

kernelstub (2.0.6) artful; urgency=medium

  * Fix adding stub to NVRAM (-s flag)
  * Re-add log file
  * Use conffiles for marking old files removed
  * Removed irrelevant ERROR output when only a single kernel is installed

 -- Ian Santopietro <ian@system76.com>  Wed, 14 Feb 2018 10:17:57 -0700

kernelstub (2.0.5) artful; urgency=medium

  * Add options for:
    - Specifying Root filesystem path (for recovery)
    - Printing current configuration options
    - Turning off setting up the loader config
  * Update automatic hooks to use / symlinks

 -- Ian Santopietro <ian@system76.com>  Mon, 05 Feb 2018 10:36:13 -0700

kernelstub (2.0.4) artful; urgency=medium

  * Use /vmlinuz and /initrd.img symlinks

 -- Ian Santopietro <ian@system76.com>  Thu, 01 Feb 2018 09:47:27 -0700

kernelstub (2.0.3) artful; urgency=medium

  * Update to include quotes in update hooks

 -- Ian Santopietro <ian@system76.com>  Tue, 30 Jan 2018 14:51:17 -0700

kernelstub (2.0.2) artful; urgency=medium

  * Update iniramfs and kernel triggers with new option names. 

 -- Ian Santopietro <ian@system76.com>  Mon, 29 Jan 2018 10:27:49 -0700

kernelstub (2.0.1) artful; urgency=medium

  * Add pop-boot package for awesome installs.

 -- Ian Santopietro <ian@system76.com>  Wed, 24 Jan 2018 11:00:12 -0700

kernelstub (2.0.0) artful; urgency=medium

  * Re-written with new, more organized sturcture.

 -- Ian Santopietro <ian@system76.com>  Wed, 24 Jan 2018 10:59:28 -0700

kernelstub (1.0.0) artful; urgency=medium

  * 1.0 release.
  * New ISC license instead of BSD.
  * First release from Git.

 -- Ian Santopietro <ian@system76.com>  Wed, 08 Nov 2017 09:47:42 -0700

kernelstub (0.19) zesty; urgency=medium

  * Improve reliability on old-kernel removal

 -- Ian Santopietro <ian@system76.com>  Tue, 25 Apr 2017 15:08:26 -0600

kernelstub (0.18) zesty; urgency=low

  * source package automatically created by stdeb 0.8.5

 -- Ian Santopietro <isantop@gmail.com>  Mon, 27 Mar 2017 11:06:36 -0600


kernelstub.py
#!/usr/bin/python3

"""
 kernelstub Version 0.2

 The automatic manager for using the Linux Kernel EFI Stub to boot
 
 Copyright 2017 Ian Santopietro <isantop@gmail.com>

 Redistribution and use in source and binary forms, with or without 
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice, this 
 list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice, 
 this list of conditions and the following disclaimer in the documentation and/
 or other materials provided with the distribution.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
 ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
 LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
 CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 INTERRUPTION) HOWEVER  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 POSSIBILITY OF SUCH DAMAGE.

 This program will automatically keep a copy of the Linux Kernel image and your
 initrd.img located on your EFI System Partition (ESP) on an EFI-compatible 
 system. The benefits of this approach include being able to boot Linux directly
 and bypass the GRUB bootloader in many cases, which can save 4-6 seconds during
 boot on a system with a fast drive. Please note that there are no guarantees 
 about boot time saved using this method. 
 
 For maximum safety, kernelstub recommends leaving an entry for GRUB in your 
 system's ESP and NVRAM configuration, as GRUB allows you to modify boot
 parameters per-boot, which loading the kernel directly cannot do. The only 
 other way to do this is to use an EFI shell to manually load the kernel and 
 give it parameters from the EFI Shell. You can do this by using:
 
 fs0:> vmlinuz-image initrd=EFI/path/to/initrd/stored/on/esp options
 
 kernelstub will load parameters from the /etc/default/kernelstub config file.
"""

import os, shutil
import logging
import argparse
import subprocess
import lsb_release

def run_command(command, simulate): # Run or simulate a command
    logging.debug("Running command: "  + command)
    if simulate == True:
        output = "Simulating: " + command
        return output
    else:
        return subprocess.getoutput(command)

def get_maj(path): # Get the major device number for path's filesystem
    dev = os.stat(path).st_dev
    major = os.major(dev)
    return major

def get_min(path): #get the minor device number for path's filesystem
    dev = os.stat(path).st_dev
    minor = os.minor(dev)
    return minor

def _parse_proc_partitions(): # Parse Partition names
    res = {}
    f = open("/proc/partitions", "r")
    for line in f:
        fields = line.split()
        try:
            tmaj = int(fields[0])
            tmin = int(fields[1])
            name = fields[3]
            res[(tmaj, tmin)] = name
        except:
            # just ignore parse errors in header/separator lines
            pass
    f.close()
    return res

def get_drive_name(path): # Get a drive's /dev name, eg sda
    major = get_maj(path)
    dInfo = _parse_proc_partitions()
    name = dInfo[(major, 0)]
    return name

def get_part_name(path): # get a partition's /dev name, eg sda1
    major = get_maj(path)
    minor = get_min(path)
    pInfo = _parse_proc_partitions()
    name = pInfo[(major, minor)]
    return name

def get_uuid(fs): # Get a UUID from a filesystem dev name
    command = "blkid /dev/" + fs
    blockId = run_command(command, False)
    blockList = blockId.split('\"')
    listIndex = 0
    for i in blockList:
        if str(i)[-6:-1] == " UUID":
            index = listIndex + 1
            listIndex = listIndex + 1
        else:
            listIndex = listIndex + 1
    fsUuid = str(blockList[index])
    return fsUuid

def get_os_name(lsbDict): # Get name of the current OS, eg 'Ubuntu', or None
    osName = lsbDict.get('ID', None)
    return osName

def get_os_version(lsbDict): # Get the version number, eg '16.10', or None
    osVersion = lsbDict.get('RELEASE', None)
    return osVersion

def find_os_entry(nvram, osLabel): # Locate an existing NVRAM entry
    findIndex = 0
    for i in nvram:
            if osLabel in i:
                    logging.info("Found OS Entry")
                    logging.info("OS Entry is:       " + nvram[findIndex])
                    return findIndex
            else:
                    findIndex = findIndex + 1
    return -1

def del_boot_entry(index, sim): # Delete an entry from the NVRAM
    command = "efibootmgr -B -b " + index
    return run_command(command, sim)

# Add an entry to the NVRAM with the specified opts
def add_boot_entry(device, 
                   partition, 
                   label, 
                   loader, 
                   root, 
                   initrd, 
                   cmdline, 
                   sim): 
    command = ('efibootmgr -d ' + device + ' -p ' + partition + ' -c -L "' + 
               label + '" -l ' + loader + ' -u "root=UUID=' + root + 
               ' initrd=' + initrd + ' ro ' + cmdline + '"')
    return run_command(command, sim)

def get_file_path(path, search): # Get path to file string, for copying stuff
    command = "ls " + path + search + "* | tail -1"
    return run_command(command, False)

def copy_files(src, dest, simulate): # Copy file src into dest
    if simulate == True:
        copy = ("Simulate copying " + src + " into " + dest)
        logging.info(copy)
        return True
    else:
        copy = ("Copying " + src + " into " + dest)
        logging.info(copy)
        try:
            shutil.copy(src, dest)
            return True
        except:
            logging.error("Copy failed! Things may not work...")
            raise FileOpsError("Could not copy one or more files.")
            return False

def ensure_dir(file_path): # Make sure a file exists, and make it if it doesn't
    directory = os.path.dirname(file_path)
    if not os.path.exists(directory):
        os.makedirs(directory)

def check_config(path): # check if the config file exists, read it
    if os.path.isfile(path) == True:
        f = open(path, "r")
        opts = f.readline()
        f.close()
        return opts
    else:
        raise ConfigError("Missing configuration, cannot continue")
        return "InvalidConfig"

def main(): # Do the thing
    # Set up argument processing
    parser = argparse.ArgumentParser(
        description = "Automatic Kernel EFIstub manager")
    parser.add_argument("-c",
                        "--cmdline",
                        help = ("Specify the kernel boot options to use (eg. "
                                "quiet splash). Default is to read from the "
                                "config file in /etc/default/kernelstub. "
                                "Options MUST be specified either in the "
                                "config file or using this option, otherwise "
                                "no action will be taken!"))
    parser.add_argument("-l",
                        "--log", 
                        help = ("Path to the log file. Default is "
                                "/var/log/kernelstub.log"))
    parser.add_argument("-lv",
                        "--log-level",
                        help = ("Sets the information level for the log file. "
                                "Default is INFO. Valid options are DEBUG, "
                                "INFO, WARNING, ERROR, and CRITICAL."))
    parser.add_argument("-k",
                        "--kernelpath",
                        help = ("Manually specify the path to the kernel image."
                                " This is useful if kernelstub can't find your "
                                "kernel automatically."))
    parser.add_argument("-i",
                        "--initrdpath",
                        help = ("Manually specify the path to the initrd image."
                                " Similar as for -k."))
    parser.add_argument("-v",
                        "--verbose",
                        action="store_true",
                        help = ("Displays extra information about the actions "
                                "being performed. "))
    parser.add_argument("-s",
                        "--simulate",
                        action="store_true",
                        help = ("Don't actually run any commands, just "
                                "simulate them. This is useful for testing."))
    args = parser.parse_args() 
    
    # Set up logging
    if args.log_level:
        file_level = getattr(logging, args.log_level.upper(), None)
        if not isinstance(file_level, int):
            raise ValueError('Invalid log level: %s' % args.log_level)
    else:
        file_level = "INFO"
    
    if args.verbose == True:
        console_level = "INFO"
    else:
        console_level = "WARNING"
    
    if args.log:
            logFile = args.log
    else:
        logFile = "/var/log/kernelstub.log"
    logging.basicConfig(level = file_level,
                        format = ('%(asctime)s %(name)-12s '
                                  '"%(levelname)-8s %(message)s'),
                        datefmt = '%m-%d %H:%M',
                        filename = logFile,
                        filemode = 'w')
    console = logging.StreamHandler()
    console.setLevel(console_level)
    formatter = logging.Formatter('%(name)-12s: %(levelname)-8s %(message)s')
    console.setFormatter(formatter)
    logging.getLogger('').addHandler(console)

    # Figure out runtime options
    if args.simulate:
        noRun = True
    else:
        noRun = False
    
    if args.kernelpath:
        linuxPath = args.kernelpath
    else:
        #logging.info("No kernel specified, attempting automatic discovery")
        linuxPath = get_file_path("/boot/", "vmlinuz")
    
    if args.initrdpath:
        initrdPath = args.initrdpath
    else:
        #logging.info("No initrd specified, attempting automatic discovery")
        initrdPath = get_file_path("/boot/", "initrd.img")
    
    # First check for kernel parameters. Without them, stop and fail
    if not args.cmdline:
        configPath = "/etc/default/kernelstub"
        try:
            kernelOpts = check_config(configPath)
        except ConfigError:
            error = ("cmdline was 'InvalidConfig'\n\n"
                     "This probably means that the config file doesn't exist "
                     "and you didn't specify any boot options on the command "
                     "line.\n"
                     "Create a new config file in /etc/default/kernelstub with "
                     "your required kernel parameters and rerun kernelstub "
                     "again!\n\n")
            logging.critical(error)
            raise CmdLineError("No Kernel Parameters found")
            return 3
    else:
        kernelOpts = args.cmdline
    
    # Drive and partition information
    driveName = get_drive_name("/")
    rootFs = get_part_name("/")
    espFs = get_part_name("/boot/efi")
    espNum = espFs[-1]
    rootUuid = get_uuid(rootFs)
    
    # OS information (from LSB)
    lsbDict = lsb_release.get_distro_information()
    osName = get_os_name(lsbDict)
    osVer = get_os_version(lsbDict)
    osLabel = osName + " " + osVer
    
    # NVRAM information
    nvram = run_command("efibootmgr", False).split("\n")
    entryIndex = find_os_entry(nvram, osLabel)
    orderNum = str(nvram[entryIndex])[4:8]
    
    # Directory Information
    osDirName = osName + "-kernelstub/"
    workDir = "/boot/efi/EFI/" + osDirName
    linuxName = "linux64.efi"
    linuxDest = workDir + linuxName
    initrdName = "initrd.img"
    initrdDest = workDir + initrdName
    
    ensure_dir(workDir) # Make sure the destination exists on the ESP

    # Log some helpful information, to file and optionally console
    logging.info("NVRAM entry index: " + str(entryIndex))
    logging.info("Boot Number:       " + orderNum)
    logging.info("Drive name is      " + driveName)
    logging.info("Root FS is on      " + rootFs)
    logging.info("ESP data is on     " + espFs)
    logging.info("ESP partition #:   " + espNum)
    logging.info("Root FS UUID is:   " + rootUuid)
    logging.info("OS running is:     " + osName + " " + osVer)
    logging.info("Kernel Params:     " + kernelOpts)
    
    # Do stuff
    logging.info("Now running the commands\n")
    try:
        copy_files(linuxPath, linuxDest, noRun)
    except FileOpsError:
        error = ("Could not copy the Kernel Image  into the ESP! This "
                 "indicates a very bad problem and it is unsafe to continue. "
                 "Aborting now...")
        logging.critical(error)
        return 2
    
    try:
        copy_files(initrdPath, initrdDest, noRun)
    except FileOpsError:
        error = ("Could not copy the initrd.img  into the ESP! This indicates "
                 "a very bad problem and it is unsafe to continue. Aborting "
                 "now...")
        logging.critical(error)
        return 3
    
    # Check for an existing NVRAM entry and remove it if present
    if entryIndex >= 0:
        logging.info("Deleting old boot entry")
        logging.info("New NVRAM:\n\n" + del_boot_entry(orderNum, noRun) + "\n")
    else:
        logging.info("No old entry to remove, skipping.")
    logging.info("New NVRAM:\n\n" + add_boot_entry("/dev/" + driveName, 
                                                   espNum, 
                                                   osLabel, 
                                                   "/EFI/" + osDirName + linuxName,
                                                   rootUuid, 
                                                   "EFI/" + osDirName + initrdName,
                                                   kernelOpts,
                                                   noRun) + "\n")
    
    try:
        copy_files("/proc/cmdline", workDir + "cmdline.txt", noRun)
    except FileOpsError:
        error = ("Could not copy the current Kernel Command line into the ESP. "
                 "You should manually copy the contents of /proc/cmdline into "
                 "the ESP to ensure you can get to it in an emergency. This is "
                 "a non-critical error, so continuing without it.")
        logging.warning(error)
        pass
    return 0

if __name__ == '__main__':
    main()
